File: K47 User Manual/7_miscellany/35_smokeSensor/Code/C/smokeSensor.c
#include <wiringPi.h>
#include <stdio.h>

typedef unsigned char uchar;
typedef unsigned int  uint;

#define     ADC_CS    0
#define     ADC_CLK   1
#define     ADC_DIO   2
#define		SensorDoPin	3
#define     LedPin      4


uchar get_ADC_Result(void)
{
	
	uchar i;
	uchar dat1=0, dat2=0;

	digitalWrite(ADC_CS, 0);

	digitalWrite(ADC_CLK,0);
	digitalWrite(ADC_DIO,1);	delayMicroseconds(2);
	digitalWrite(ADC_CLK,1);	delayMicroseconds(2);
	digitalWrite(ADC_CLK,0);

	digitalWrite(ADC_DIO,1);    delayMicroseconds(2); 
	digitalWrite(ADC_CLK,1);	delayMicroseconds(2);
	digitalWrite(ADC_CLK,0);

	digitalWrite(ADC_DIO,0);	delayMicroseconds(2); 
	
	digitalWrite(ADC_CLK,1);	
	digitalWrite(ADC_DIO,1);    delayMicroseconds(2);
	digitalWrite(ADC_CLK,0);	
	digitalWrite(ADC_DIO,1);    delayMicroseconds(2);
	
	for(i=0;i<8;i++)
	{
		digitalWrite(ADC_CLK,1);	delayMicroseconds(2);
		digitalWrite(ADC_CLK,0);    delayMicroseconds(2);

		pinMode(ADC_DIO, INPUT);
		dat1=dat1<<1 | digitalRead(ADC_DIO);
	}
	
	for(i=0;i<8;i++)
	{
		dat2 = dat2 | ((uchar)(digitalRead(ADC_DIO))<<i);
		digitalWrite(ADC_CLK,1); 	delayMicroseconds(2);
		digitalWrite(ADC_CLK,0);    delayMicroseconds(2);
	}

	digitalWrite(ADC_CS,1);

	pinMode(ADC_DIO, OUTPUT);
	
	return(dat1==dat2) ? dat1 : 0;
}

int main(void)
{
	uchar analogVal;

	if(wiringPiSetup() == -1)
	{
		printf("setup wiringPi failed !\n");
		return -1; 
	}

	pinMode(ADC_CS,  OUTPUT);
	pinMode(ADC_CLK, OUTPUT);
	pinMode(LedPin,  OUTPUT);
	pinMode(SensorDoPin, INPUT);

	while(1)
	{
		analogVal = get_ADC_Result();
		printf("%d\n", analogVal);

		if(!digitalRead(SensorDoPin))
		{
			digitalWrite(LedPin, LOW);
			delay(100);
			digitalWrite(LedPin, HIGH);
			printf("\n***********	\
				\n  Danger!    \
				\n***********  \
				\n");
			delay(1000);
		}
		else
		{
			digitalWrite(LedPin, LOW);
		}
		delay(100);
	}

	return 0;
}

File: K47 User Manual/7_miscellany/35_smokeSensor/Code/Python/smokeSensor.py
#!/usr/bin/env python

#
# This is a program for MQ-2 Gas Sensor Module.
# It could detect danger gas and smokes.
# This program depends on ADC0832 ADC chip. Follow 
# the instruction book to connect the module and 
# ADC0832 to your Raspberry Pi.
#


import RPi.GPIO as GPIO
import ADC0832
import time

SensorDoPin = 15
LedPin = 16					


def setup():
	GPIO.setmode(GPIO.BOARD)		# Numbers GPIOs by physical location
	GPIO.setup(LedPin, GPIO.OUT)		
	GPIO.setup(SensorDoPin, GPIO.IN)
	ADC0832.setup()					

def loop():
    while True:
	    analogVal = ADC0832.getResult(0)	# Get analog value from ADC0832
	    print analogVal					# Print analog value
							
	    if not GPIO.input(SensorDoPin):						
                print '    ****************'
		print '    * ! DANGER ! *'
		print '    ****************'
		print ''
				
		GPIO.output(LedPin, GPIO.LOW)	
		time.sleep(0.1)
		GPIO.output(LedPin, GPIO.HIGH)
			
	    else:
	        GPIO.output(LedPin, GPIO.LOW)
	    time.sleep(1)					

def destroy():
	GPIO.cleanup()				# Release resource

if __name__ == '__main__':		# Program start from here
	setup()
	try:
		loop()
	except KeyboardInterrupt:
		destroy()
